perf(api, dispatch-service): switch storage backend to Google Cloud storage bucket

bilalshaikh42 opened the pull request.

bilalshaikh42 committed:
revert:  "refactor(dispatch-service): moved hsload to complete processor"

This reverts commit 32116df13dee7de487a2a5e0ed2afaadde303677.

bilalshaikh42 committed:
perf(api,dispatch-service): change storage backend to Google Cloud

change the storage bucket used in the backed from the uconn hpc to the google cloud storage buckets.

closes #4039

bilalshaikh42 commented on apps/dispatch-service/src/file/file.service.ts:
```diff
@@ -82,47 +83,32 @@ export class FileService {
+                .getFileInfo(id, file.location.path)
+                .pipe(
+                  map((fileInfo: FileInfo): ProjectFileInput => {
+                    // TODO do we want to throw here if we don't have a size?
```
@jonrkarr You were throwing an error here if we were unable to get the file size. Is this something that we should do here? I am currently just setting the size to undefined if we are not able to get it, as it feels silly to fail the whole run just because one of the file sizes is missing. 
What do you think?

bilalshaikh42 committed:
refactor(dispatch-service): split thumnail processing and file processing

jonrkarr commented:
Not throwing an error is fine. Previously i believe the size was set to zero if it couldn't be obtained. Better would be undefined. The UI is already set up to display undefined as N/A.

bilalshaikh42 committed:
docs: add developer docs about configuring endpoints

bilalshaikh42 committed:
refactor: remove filepaths from common config

front end apps should use the API to access specific files on the storage buckets to prevent changes to location breaking existing runs

closes #3981

bilalshaikh42 committed:
refactor: front end apps should not use external endpoints

the front end apps do not need to specify the external endpoint since they are always accessing public apis.
for example, if running locally, they will always be able to access the localaly deployed urls. If deployed, they will not be using the "local" env
The additional case of using kubernetes local dns names also does not apply to front end apps

bilalshaikh42 committed:
feat: add cors and lifecycle config for storage buckets

add python scripts to set CORS setting for google stroage buckets and lifecycle rules for temp
storage bucket

closes #3994

bilalshaikh42 committed:
refactor(config): change thumbnail type to enum

bilalshaikh42 committed:
refactor(config): remove file paths from shared config to storage

information about paths on the s3 bucket is now in the shared config library. This means only backend apps should be dealing with storage bucket organization

closes #3981

bilalshaikh42 committed:
feat(api): add endpoint to download files and thumbnails

files can now be downloaded from the api through the /files/runID/fileID/download endpoint.
clients no longer need to get the file metadata and then follow the url

adresses #3981

bilalshaikh42 committed:
chore: cleanup

bilalshaikh42 committed:
refactor(storage): enable ssl validation

we had previously disabled ssl validation due to limitations of on prem storage. certificate
valiation is now enabled

closes #4059

bilalshaikh42 committed:
chore: spelling fixes

bilalshaikh42 commented:
@jonrkarr still needs to add/update tests and do a final end to end check, but everything should be good for review

bilalshaikh42 committed:
refactor: move thumbnail information to datamodel

move thumbnail types to shared datamodel
remove ineffective scope rules (to be replaced)

bilalshaikh42 committed:
feat(api): add endpoint for putting endpoints of files.

removes the need for API or frontends to calculate files paths for any files. This keeps the storage paths restrcited to the dispatch-service.

bilalshaikh42 committed:
refactor(dispatch-service): move fetching manifest into common service

bilalshaikh42 committed:
chore: lint and test fixes

biosimulations-daemon commented:
:tada: This PR is included in version 9.0.0 :tada:

The release is available on [GitHub release](https://github.com/biosimulations/biosimulations/releases/tag/v9.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
